File: algorithms/leetcode/098-validate-binary-search-tree/validate-binary-search-tree.js
const isValidBST = root => {
    const validate = (node, min, max) => {
        if (!node) {
            return true;
        }

        if (
            (node.val > min && node.val < max) &&
            ((node.left === null) || (node.left.val < node.val)) &&
            ((node.right === null) || (node.right.val > node.val))
        ) {
            return validate(node.left, min, node.val) && validate(node.right, node.val, max);
        } else {
            return false;
        }
    };

    return validate(root, Number.NEGATIVE_INFINITY, Number.POSITIVE_INFINITY);
};

File: algorithms/leetcode/098-validate-binary-search-tree/ValidateBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private boolean isValidBST(TreeNode root, Integer lowerLimit, Integer upperLimit) {
        if (root == null) {
            return true;
        }
        
        if (lowerLimit != null && root.val <= lowerLimit || upperLimit != null && root.val >= upperLimit) {
            return false;
        }
        
        return isValidBST(root.left, lowerLimit, root.val) && isValidBST(root.right, root.val, upperLimit);
    }
    
    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, null, null);
    }
}
